importer/taric.py

```
@@ -124,7 +125,12 @@ def save(
 
         for message_data in data["message"]:
             self.message.save(message_data, transaction.id)
-        self.parent.workbasket.clean()
+
+        try:
+            transaction.clean()
+        except ValidationError as e:
+            for message in e.messages:
+                logger.error(message)
```

Surely we should raise the error as well. No reason to suppress it.

---

The reasoning is that we don't actually want to halt the import job. I guess we could handle it upwards instead? But ultimately  we won't be able to keep prod up to date as a result of our business rule bugs, whereas at least this way we can have data in the system that has been validated elsewhere whilst we fix the bugs.

---

It may have been a good idea for me to read the PR summary!

So if we're assuming it is a bug on our side this makes sense. But I'm not convinced we should always be thinking that (plenty of times we've found silly data from those providing the data - admittedly we've had to adapt to it most times anyway). And once the data is in the system it's a pain to revert an import (not impossible, but a pain).

---

Yeah I do agree, it's clearly will not always be a bug necessarily, but if we're bringing data in via this route it will basically either be EU data that we pretty much have to accept verbatim, or manually prepared data. In either case (unless something fails as the first element) we're always going to have to handle reverting the previous data that has been loaded if we want to go back? The only real difference is that this way you a) have the option of not reverting it, if you're sure it's what you want depsite the errors, and you b) have all the data in the database so you can do a more thorough analysis.

I was going to suggest if there's somewhere we can put a flag to toggle behaviour, but then I'm not sure if the explosion behaviour is desired in any environment due to a). Is there instead someway we could signal that an import job contains errors without making it impossible to proceed? Because the thing I would be worried about is not having any feedback from the import UI that something threw an error.

---

What we do get right now if an error occurs is sentry telling us what is going on. I'm not convinced we get that with this change. And I don't think the job will be marked as errorred either. So there is some difficulty here (may not realise we have an error until months after the fact), finding the exact error messages etc.

I can see the reasoning behind it. I'm just wondering if this will in fact make it harder to detect problems (despite it being easier to investigate once we know we have one).

---

I agree that we shouldn't be making it harder to detect problems… can you see any ways of achieving that? Could we catch the errors and then rethrow them all at the end of the job?

---

We could use a different logger for this, that routes messages to the ordinary handler, as well as a handler that collects and outputs the errors for this.

---

Here try this https://docs.sentry.io/platforms/python/usage/#capturing-errors

---

So if I did:

    import settings
    if settings.SENTRY_ENABLED:
        from sentry_sdk import capture_exception

And subsequently:

            transaction.clean()
        except ValidationError as e:
            for message in e.messages:
                logger.error(message)
            if settings.SENTRY_ENABLED:
                capture_exception(e)

Does that integrate it in the way we would like/be comfortable with?